### PR TITLE
Adds scoped slots feature.

### PR DESCRIPTION
**Pull request by dclause**
I wish to suggest this PR.

It allows to customize component using there scoped slots. Somehow using the current API, you can customize a component via their named slots. Some components thought, let's you create complex and interesting features such as QTree or QTable. But to do so, you need to use the props injected into their scoped slots. For instance : 
- https://quasar.dev/vue-components/tree#customize-content
- https://quasar.dev/vue-components/table#body-slots

Here are some examples created using Nicegui and this PR: 

image

Some notes about the API: 
- Slot class has been extended to receive an optional template. A Slot with a template will be rendered as a scoped slot.
- The rest of the API has not changed, so it is transparent to anything else.
- Scoped slots uses props injected into the slot. The injected data is always named **props** and can be accessed from the vue template.
- There is no way to make that pure python that I could think of for technical reasons (I guess its kinda obvious but I can expand on this). I would argue that:
  - scoped slots can only be discovered using the Quasar documentation, hence it is an advanced use case
  - once discovered in the documentation, you have a vue example that you simply have to past in your own code. The only change will be to rename the injected props to props on some cases. 

I am looking forward to have your feedback on this :) I strongly believe this is a nice addition to the API because that is the only way to avoid creating custom components and custom js files in our projects to create stuffs like the table above.

### CODE REVIEW

File: nicegui/templates/index.html [hunk 1]
```diff
@@ -70,12 +70,24 @@
...
+            const rendered = [];
+            if (data.template) {
+              rendered.push(Vue.h({
+                props: { props: { type: Object, default: {} } },
+                template: data.template,
+              }, {
+                props: props
+              }));
+            }
+            const ids = ('ids' in data) ? data.ids : data;
```
Comment (falkoschindler) [inline nicegui/templates/index.html:84]: Do we really need the second case where `ids = data`? Shouldn't there always be `data.ids`?

**Review by falkoschindler — APPROVED** [review]
Very impressive work, @dclause! Thanks a lot!

I reviewed your code, committed just a few tiny adjustments here and there (mostly whitespace etc.), and commented only one question.

The only tiny downside of this solution is that one needs to write Vue templates as multiline strings without support from the Python world. But I guess it would be immensely complicated to allow creating these templates with NiceGUI elements. So I think this is a very good compromise for such an advanced feature which will mainly be used from experienced developers.

File: nicegui/templates/index.html (hunk 1 above)
Comment (dclause) [reply nicegui/templates/index.html]: Nice catch ! Indeed, line 84 is no more needed. I add many back and forth on this trying things and that was my first attempt when slots and scoped_slots where in separated structure. Anyway: here is done a commit to clean that.

**Comment by dclause** [comment]
I understand your concern with vue templates in python code. That is indeed a downside, but 
- the feature is advanced enough to give that to more "poweruser"
- I could not find any technical way to do this and render the scoped slots using the jsx / render syntax
- I don't see any technical way to get the slots arguments on the python side when coding because at that step, nothing exists: no vue.js started yet, no element created, nothing. I don't known how we can do that

However, now the feature exists using vue templating, nothing prevents to make it better in the future if we find-out how.

**Comment by falkoschindler** [comment]
@dclause I totally agree with you. 👍

**Comment by rodja** [comment]
I think it is almost ready to merge. Maybe you could add a automated test or two for the scoped slots @dclause? Also some documentation would be nice. But the "normal" slots are also not well documented. I would be fine to defer this to a separate task.

**Comment by dclause** [comment]
@rodja Testing slots named and scoped would require a component that has both. qtable in PR #500 is elligible to that. Can we make a unique test_qtable.py file to test both the qtable and the usage of it's slots ? Just like per the example provided in the PR #500 but with additional asserts ?

Also, I could not figure out how to run the tests. Should I use the docker or my local machine ? Locally, I guess I succeeded in installing everything but I still got this : 
` > pytest tests/test_upload.py`
```
ImportError while loading conftest 'C:\Users\xxx\Projets\nicegui\tests\conftest.py'.
tests\conftest.py:9: in <module>
    from nicegui import Client, globals
nicegui\__init__.py:1: in <module>
    from . import elements, globals, ui
nicegui\ui.py:60: in <module>
    from .elements.upload import Upload as upload
nicegui\elements\upload.py:7: in <module>
    from ..nicegui import app
nicegui\nicegui.py:27: in <module>
    socket_manager = SocketManager(app=app, mount_location='/_nicegui_ws/', json=json)
E   TypeError: SocketManager.__init__() got an unexpected keyword argument 'json'
```

Can I have help on this ?

**Comment by falkoschindler** [comment]
@dclause Using qtable for the tests is certainly a good idea. But can't we also use `ui.input` for named slots and `ui.tree` for scoped ones? Then this wouldn't depend on #500.

The error with `SocketManager` looks like you need to update your fastapi-socketio installation. We need version 0.0.10 for being able to pass the keyword argument `json`.

**Comment by rodja** [comment]
> Can we make a unique test_qtable.py file to test both the qtable and the usage of it's slots

Yes, that sounds great. I think we can merge this and then add the tests in another pull request. Regarding the tests: your error looks strange and definitely not jet covered by https://github.com/zauberzeug/nicegui/blob/main/tests/README.md. Maybe something related to the orjson wrapping? Does the NiceGUI examples starts up locally? I mean with your local installation like described in https://github.com/zauberzeug/nicegui/blob/main/CONTRIBUTING.md#setup.

**Comment by dclause** [comment]
Let's merge this then.
Regarding tests, no, I was not able to run locally the project. I have an error when running `python main.py`:
`TypeError: deprecated() got an unexpected keyword argument 'name'``

I have run the rpoject successfully through the docker though for programming. 
I am not able to run the examples using a local nicegui install too. That means the example using the new API addition could not be tested without some trickery on my side to get the nicegui to point to itself.

I am surely missing something here. I am fully willing to create the table tests, but without understanding why I can't run them, that is a bit hard here.

**Comment by rodja** [comment]
> I am not able to run the examples using a local nicegui install too.

@rbeeli What is your output of 

```bash
python3 -m pip install -e .
```

**Comment by dclause** [comment]
1. I have created a new venv and tried the install fresh from there:
`pip install -e .`
then run the main.: `python main.py` or `python -m nicegui`
I add errors because two packages where missing: `itsdangerous` and `docutils` I add to install them manually. 
Once done, I can run nicegui properly locally.


2. Then I installed the tests requirements: 
`pip install -r tests/requirements.txt`
And could then run the tests properly from the main package: 
`pytest tests/test_progress.py`

3. I could also run the example from the main folder: 
` python examples/table_and_slots/main.py`
NOTE: this examples exists only in the PR #500 branch which proves it to work locally.

In conclusion: I get the setup working properly now. I will expand the PR #500 with a test for table and slots. In the meantime, could you review that PR regarding the latest correction done ?

Finally, I would say we can also close the topic here: thanks for merging :)
